lib/cretonne/src/ssa.rs

```diff
@@ -0,0 +1,716 @@
+/// struct Variable(u32);
+/// impl EntityRef for Variable {
+///     fn new(index: usize) -> Self {
+///         assert!(index < (u32::MAX as usize));
+///         Variable(index as u32)
+///     }
+///
+///     fn index(self) -> usize {
+///         self.0 as usize
+///     }
+/// }
+///
+/// let mut func = Function::new();
+/// let mut ssa: SSABuilder<Variable> = SSABuilder::new();
+/// let ebb0 = func.dfg.make_ebb();
+/// let ebb1 = func.dfg.make_ebb();
+/// let cur = &mut Cursor::new(&mut func.layout);
+/// cur.insert_ebb(ebb0);
+/// cur.insert_ebb(ebb1);
+/// cur.goto_bottom(ebb0);
+///
+/// // ebb0:
+/// let block0 = ssa.declare_ebb_header_block(ebb0);
+/// let x_var = Variable(0);
+/// // x = 1
+/// let x_ssa = func.dfg.ins(cur).iconst(I32, 1);
+/// ssa.def_var(x_var, x_ssa, block0);
+/// let y_var = Variable(1);
+/// // y = 2
+/// let y_ssa = func.dfg.ins(cur).iconst(I32, 2);
+/// ssa.def_var(y_var, y_ssa, block0);
+/// let z_var = Variable(2);
+/// let x_use1 = ssa.use_var(&mut func.dfg, x_var, I32, block0);
+/// let y_use1 = ssa.use_var(&mut func.dfg, y_var, I32, block0);
+/// // z = x + y
+/// let z1_ssa = func.dfg.ins(cur).iadd(x_use1, y_use1);
+/// ssa.def_var(z_var, z1_ssa, block0);
+/// let y_use2 = ssa.use_var(&mut func.dfg, y_var, I32, block0);
+/// // brnz v1, ebb1
+/// let jump_inst: Inst = func.dfg.ins(cur).brnz(y_use2, ebb1, &[]);
+/// // The first basic block is now filled. Because it has no predecessors, we can seal it.
+/// ssa.seal_ebb_header_block(ebb0, &mut func.dfg);
+///
+/// // Here we change of basic block because of the branch instruction.
+/// let block1 = ssa.declare_ebb_body_block(ebb0, block0);
+/// let x_use2 = ssa.use_var(&mut func.dfg, x_var, I32, block1);
+/// let z_use1 = ssa.use_var(&mut func.dfg, z_var, I32, block1);
+/// // z = x + z
+/// let z2_ssa = func.dfg.ins(cur).iadd(x_use2, z_use1);
+/// ssa.def_var(z_var, z2_ssa, block1);
+///
+/// // ebb1:
+/// let block2 = ssa.declare_ebb_header_block(ebb1);
+/// ssa.declare_ebb_predecessor(ebb1, block0, jump_inst);
+/// cur.goto_bottom(ebb1);
+///
+/// let x_use3 = ssa.use_var(&mut func.dfg, x_var, I32, block2);
+/// let y_use3 = ssa.use_var(&mut func.dfg, y_var, I32, block2);
+/// // y = x + y
+/// let y2_ssa = func.dfg.ins(cur).iadd(x_use3, y_use3);
+/// ssa.def_var(y_var, y2_ssa, block2);
+///
+/// // All the predecessors of ebb1 have been declared, we can seal it.
+/// ssa.seal_ebb_header_block(ebb1, &mut func.dfg);
+/// ```
+///
+/// At this point `ebb1` has two value arguments and the branch instruction uses the correct
+/// SSA values to feed to `ebb1`.
+
+impl<Variable> SSABuilder<Variable>
+    where Variable: EntityRef + Hash
+{
+    /// Declares a new definition of a variable in a given basic block.
+    /// The SSA value is passed as an argument because it should be created with
+    /// `ir::DataFlowGraph::append_result`.
+    pub fn def_var(&mut self, var: Variable, val: Value, block: Block) {
+        if match self.variables.get_mut(var) {
```

Comment:
Ok, I abandonned the `SparseMap` and switched to an `EntityMap` which makes more sense, as long as `Variable` implementes `Default`.